### PR TITLE
sql-query output encrypted

### PR DESCRIPTION
**Pull request by BumbleFeng**
Merge after https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/13633

Signed-off-by: Feng Huang <fehuang@redhat.com>

### CODE REVIEW

File: reconcile/sql_query.py [hunk 1]
```diff
@@ -176,9 +197,32 @@ def collect_queries(query_name=None):
         # Output can be:
         # - stdout
         # - filesystem
+        # - encrypted
         output = sql_query['output']
         if output is None:
             output = 'stdout'
+        elif output == 'encrypted':
+            requestor = sql_query['requestor']
```
Comment (maorfr) [inline reconcile/sql_query.py:205]: we should use `.get` when we are uncertain if the object contains the key

File: reconcile/sql_query.py [hunk 2]
```diff
@@ -176,9 +197,32 @@ def collect_queries(query_name=None):
...
+            if requestor is None:
+                logging.error(
+                    'a requestor is required to get encrypted output'
+                )
+                sys.exit(ExitCodes.ERROR)
+            public_gpg_key = requestor['public_gpg_key']
```
Comment (maorfr) [inline reconcile/sql_query.py:211]: same as above

File: reconcile/sql_query.py [hunk 3]
```diff
@@ -176,9 +197,32 @@ def collect_queries(query_name=None):
...
+            user_name = requestor['org_username']
+            if public_gpg_key is None:
+                logging.error(
+                    ['user %s does not have a public gpg key'],
+                    user_name
+                )
+                sys.exit(ExitCodes.ERROR)
+            ok, error_message = gpg_key_valid(public_gpg_key)
```
Comment (maorfr) [inline reconcile/sql_query.py:219]: the gpg key will be validated by the `user-validator` integration

File: reconcile/sql_query.py [hunk 4]
```diff
@@ -278,14 +332,31 @@ def filesystem_closing_message():
         'echo;',
         'echo Get the sql-query results with:;',
         'echo;',
-        'echo  oc cp ${HOSTNAME}:/tmp/query-result.txt ',
+        'echo  oc cp ${HOSTNAME}:tmp/query-result.txt ',
```
Comment (maorfr) [inline reconcile/sql_query.py:335]: this should probably be a separate PR? :1st_place_medal:

File: reconcile/sql_query.py [hunk 5]
```diff
@@ -211,6 +255,12 @@ def collect_queries(query_name=None):
             **tf_resource_info,
         }
 
+        if output == 'encrypted':
+            smtp_client = SmtpClient(settings=settings)
+            item['recipient'] = smtp_client.get_recipient(
```
Comment (maorfr) [inline reconcile/sql_query.py:260]: :heart:

**Review by maorfr — APPROVED** [review]
LGTM

this is very nice! i like the usage of a configmap a lot.

File: reconcile/sql_query.py [hunk 6]
```diff
@@ -35,7 +37,7 @@
     spec:
       {% if PULL_SECRET is not none %}
       imagePullSecrets:
-      - name: {{ JOB_NAME }}
+      - name: {{ PULL_SECRET }}
```
Comment (apahim) [inline reconcile/sql_query.py:40]: separate commit please

File: reconcile/sql_query.py [hunk 7]
```diff
@@ -278,14 +332,31 @@ def filesystem_closing_message():
...
     ]
 
 
+def encrypted_closing_message(recipient):
+    return [
+        'cat /gpg/gpg-key | base64 --decode | ',
+        'gpg --homedir /tmp/.gnupg --import;',
+        'gpg --trust-model always --homedir /tmp/.gnupg --armor -r ',
+        recipient,
+        ' --encrypt /tmp/query-result.txt;',
```
Comment (apahim) [inline reconcile/sql_query.py:349]: Will the /tmp/query-result.txt be kept unencrypted in the container? If yes, it shouldn't. The encryption is meant to make sure only the requestor has access to the results... we should remove the plain results file.

File: reconcile/sql_query.py [hunk 8]
```diff
@@ -423,34 +523,12 @@ def run(dry_run, enable_deletion=False):
...
-                                  oc_map=oc_map,
-                                  cluster=query['cluster'],
-                                  namespace=query['namespace']['name'],
-                                  resource_type='Secret',
-                                  name=query['name'],
-                                  enable_deletion=enable_deletion)
-        except StatusCodeError:
-            LOG.exception("Error removing ['%s' '%s' 'Secret' '%s']",
-                          query['cluster'], query['namespace']['name'],
-                          query['name'])
```
Comment (apahim) [inline reconcile/sql_query.py:453]: nice refactoring. should probably go in a separate commit.

**Comment by apahim** [comment]
Looks very nice overall. Great job. A security concern with the results file and a couple of small nitpics related to commits separation.
